appliances/Harbor/appliance.sh

```diff
@@ -0,0 +1,304 @@
+HARBOR_DB_PASSWORD="${HARBOR_DB_PASSWORD:-$(gen_db_password)}" # Defaults to autogenerated password
+HARBOR_HOSTNAME="${HARBOR_HOSTNAME:-0.0.0.0}" # Listens in all interfaces by default
+HARBOR_SSL_CERT="${HARBOR_SSL_CERT:-AG}" # Defaults to "AG", indicating "AutoGenerated"
+HARBOR_SSL_KEY="${HARBOR_SSL_KEY:-AG}" # Defaults to "AG", indicating "AutoGenerated"
+HARBOR_PERSISTENT_DEV="${HARBOR_PERSISTENT_DEV:-None}" # Listens in all interfaces by default
+
+
+### Installation Stage => Installs requirements, downloads and unpacks Harbor 
+service_install() {
+    msg info "Checking internet access..."
+    check_internet_access
+    install_requirements
+    download_unpack_harbor
+    msg info "Installation phase finished"
+}
+
+### Configuration Stage => Configures Harbor YAML and SSL certificates
+service_configure() { 
+    msg info "Starting configuration..."
+    mount_persistent
+    ## SSL CERTS
+
+    mkdir /root/certs
+    if [ "$HARBOR_SSL_CERT" = "AG" ] || [ "$HARBOR_SSL_KEY" = "AG" ]; then
+        msg info "Autogenerating SSL certificates..."
+        generate_ssl_certs
+    else
+        msg info "Configuring provided SSL certificates..."
+        echo $HARBOR_SSL_CERT >> mkdir /root/certs/server.crt
+        echo $HARBOR_SSL_KEY >> mkdir /root/certs/server.key
+    fi
+
+    msg info "Configuring Harbor YAML file..."
+
+    gawk -i inplace -v cert="/root/certs/server.crt" -v key="/root/certs/server.key" -v hostname="$HARBOR_HOSTNAME" '
+    {
+        if ($1 == "certificate:") {
+            print "  certificate:", cert
+        } else if ($1 == "private_key:") {
+            print "  private_key:", key
+        } else if ($1 == "hostname:") {
+            print "hostname:", hostname
+        } else {
+            print
+        }
+    }' /root/harbor/harbor.yml
+
+    gawk -i inplace -v admin_pwd="$HARBOR_ADMIN_PASSWORD" -v db_pwd="$HARBOR_DB_PASSWORD" '
+    {
+        if ($1 == "harbor_admin_password:") {
+            print "harbor_admin_password:", admin_pwd
+        } else if ($1 == "password:") {
+            print "  password:", db_pwd
+        } else {
+            print
+        }
+    }' /root/harbor/harbor.yml
+    msg info "Configuration phase finished"
+}
+
+service_bootstrap() { # Running Harbor installer => will generate and run docker-compose
+    msg info "Starting bootstrap..."
+    /root/harbor/install.sh
+    if [ $? -ne 0 ]; then
+	    msg error "Harbor installation script failed, aborting..."
+        exit 1
+    else
+        msg info "Harbor installation script finished successfully. Waiting 60s before final health checks..."
+    fi
+    sleep 60
+    msg info "Running final health checks..."
+    wait_for_docker_containers
+    cleanup_installation
+    msg info "Bootstrap phase finished"
+
+}
+
+### Function Definitions
+
+check_internet_access() {
+    # Ping Google's public DNS server
+    if ping -c 1 8.8.8.8 &> /dev/null; then
```

Comment:
Do we know if airgapped install is considered to be a requirement?

Reply:
No airgapped needed :)